src/htcp.cc

```diff
@@ -313,6 +314,18 @@ htcpHexdump(const char *tag, const char *s, int sz)
 #endif
 }
 
+static bool
+parseUint16(char *buf, int sz, uint16_t &out, const char *field)
+{
+    if (sz < 2) {
+        debugs(31, 3, "too short for " << field);
+        return false;
+    }
+
+    out = ntohs(*(uint16_t *) buf);
```

Comment:
Please do not cast `char` buffers to integers. Integers have alignment and possibly other requirements that raw buffers may not support. Your earlier attempt with memcpy() was moving this code in the right direction.

> [Amos](https://github.com/squid-cache/squid/pull/2164#discussion_r2318841772): The original code avoided double-copy and risks of input corruption from non-atomic ntohs byte swapping.

I do not know what "double-copy" is in this context, but there are no atomicity concerns in this non-threaded code operating on local data AFAICT. Correctly copying raw bytes to an integer and then applying ntohs() function to that integer will not corrupt anything.

Reply:
https://github.com/squid-cache/squid/pull/2164/commits/af2dd78a96f1dcd4d9406da5fef89467d5ed67e9